### PR TITLE
[Cody PLG] add mutation & gql field to upgrade users to Cody Pro

### PR DESCRIPTION
**Pull request by thenamankumar**
This is temporary for Dec GA and will be removed after Feb GA eventually. 
- Adds `cody_pro_enabled_at` column to `users` table. 
- Adds `cody_pro_enabled` and `cody_pro_enabled_at` gql fields on `type User`.
- Adds `upgrade_to_cody_pro` gql mutation.

## Test plan
```graphql
mutation {
  upgradeToCodyPro(user: "VXNlcjox") {
    id
    username
    codyProEnabled
    codyProEnabledAt
  }
}
```

- `sg start dotcom`
- Set feature flat `cody_pro_dec_ga` to true
- Performing the gql mutation should be successful
- Performing the mutation again should result in "user already...."
- Set the feature flag to false.
- Performing the mutation again should result in "feature not enabled"

### CODE REVIEW

File: cmd/frontend/graphqlbackend/user.go [hunk 1]
```diff
@@ -299,6 +320,45 @@ func (r *schemaResolver) UpdateUser(ctx context.Context, args *updateUserArgs) (
...
+		return nil, errors.Errorf("this feature is not enabled")
+	}
+
+	userID, err := UnmarshalUserID(args.User)
+	if err != nil {
+		return nil, err
+	}
+
+	// 🚨 SECURITY: Only the authenticated user can update their properties.
+	if err := auth.CheckSameUser(ctx, userID); err != nil {
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/user.go:342]: The docstring of this mutation says "Only the user and site admins may perform this mutation."

but the logic here only allowed "the user"?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -6494,6 +6500,14 @@ type User implements Node & SettingsSubject & Namespace {
     Null, if not overwritten.
     """
     codeCompletionsQuotaOverride: Int
+    """
+    Whether the user has enrolled for Cody Pro.
+    """
+    codyProEnabled: Boolean!
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/schema.graphql:6506]: Yeah, this is just good to have thing.

File: cmd/frontend/graphqlbackend/user.go (hunk 1 above)
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/user.go:342]: The comment is incorrect. Will fix this.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 2 above)
[the remark this answers is not in the document]
Comment (vdavid) [reply cmd/frontend/graphqlbackend/schema.graphql:6506]: Alternatively to removing it, we could explicitly mark this as a convenience field in the comment.

**Review by vdavid — APPROVED** [review]
Nice!

**Review by chrsmith — APPROVED** [review]
Some minor nit picky things, but this all looks right to me.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 3]
```diff
@@ -90,6 +90,12 @@ type Mutation {
     """
     updateUser(user: ID!, username: String, displayName: String, avatarURL: String): User!
     """
+    Upgrades the user to Cody Pro.
+
+    Only the user and site admins may perform this mutation.
+    """
+    upgradeToCodyPro(user: ID!): User!
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/schema.graphql:97]: This is only meant for the Dec GA scope, where we want to just capture the interest.

File: internal/database/users.go [hunk 4]
```diff
@@ -616,6 +617,15 @@ func (u *userStore) Update(ctx context.Context, id int32, update UserUpdate) (er
 	return nil
 }
 
+// NOTE(naman): THIS IS TEMPORARY FOR DEC GA
+// Upgrade user to cody pro plan
+func (u *userStore) UpgradeToCodyPro(ctx context.Context, id int32) (err error) {
+	// TODO: update usage limits as well
+	query := sqlf.Sprintf("UPDATE users SET cody_pro_enabled_at=NOW() WHERE id=%d AND cody_pro_enabled_at IS NULL", id)
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply internal/database/users.go:624]: Yeah, I put that logic in the resolver, but will move it here in the store.

File: cmd/frontend/graphqlbackend/user.go [hunk 5]
```diff
@@ -299,6 +320,45 @@ func (r *schemaResolver) UpdateUser(ctx context.Context, args *updateUserArgs) (
 	return UserByIDInt32(ctx, r.db, userID)
 }
 
+type upgradeToCodyProArgs struct {
+	User graphql.ID
+}
+
+func (r *schemaResolver) UpgradeToCodyPro(ctx context.Context, args *upgradeToCodyProArgs) (*UserResolver, error) {
+	if !envvar.SourcegraphDotComMode() {
+		return nil, errors.Errorf("this feature is only available on sourcegraph.com")
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/user.go]: updated

File: cmd/frontend/graphqlbackend/user.go (hunk 1 above)
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/user.go:342]: fixed

File: internal/database/users.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (thenamankumar) [reply internal/database/users.go]: updated

**Pull request merged**